### PR TITLE
Homework 01

### PR DESCRIPTION
**Pull request by Taras-Lavriv**

### CODE REVIEW

File: Task_1/sample_project/calculator/calc.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+def check_power_of_2(a: int) -> bool:
+    return not (bool(a & (a - 1)))
```
Comment (annkupriyanova) [inline Task_1/sample_project/calculator/calc.py:2]: What does happen if a=0?

File: Task_2/tests/test_fib_checker.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    assert check_fibonacci([8, 13, 21, 34, 55, 89, 144, 233])
+
+
+def test_negative_case():
+    """Testing that not part of fibonacci give False"""
+    assert not check_fibonacci([1, 2, 4])
+
+
+def test_less_3():
+    assert check_fibonacci([2, 3])
```
Comment (annkupriyanova) [inline Task_2/tests/test_fib_checker.py:15]: Is it a fibonacci sequence?

File: Task_2/fib_checker.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+from typing import Sequence
+
+
+def generate_fib(start):
+    a, b = 0, 1
+    while True:
```
Comment (annkupriyanova) [inline Task_2/fib_checker.py:6]: It's not a good idea to use `while True` because it can potentially cause an infinite loop

File: Task_3/Min_Max_tuple.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+def find_maximum_and_minimum(file_name: str):
+    with open(file_name) as fi:
+        for lines in fi:
+            line = lines.strip()
+            list_of_data = line.split(', ')
```
Comment (annkupriyanova) [inline Task_3/Min_Max_tuple.py:5]: According to the task, a file has only one number per line, and you should read it line-by-line: "Write down the function, which reads input line-by-line, and find maximum and minimum values."

File: Task_3/Tests/test_Min_Max_tupl.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+def test_positive():
+    assert find_maximum_and_minimum(os.path.join(os.path.dirname(__file__), 'data1.txt')) == (1, 5)
+
+
+def test_one():
+    assert find_maximum_and_minimum(os.path.join(os.path.dirname(__file__), 'data2.txt')) == (256, 256)
+
+
+def test_none():
```
Comment (annkupriyanova) [inline Task_3/Tests/test_Min_Max_tupl.py:13]: There is no need to test a case with an empty file because the task says: "We guarantee, that file exists and contains line-delimited integers."

File: Task_3/Tests/test_Min_Max_tupl.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+from Homework_01.Task_3.Min_Max_tuple import find_maximum_and_minimum
+import os
+
+
+def test_positive():
```
Comment (annkupriyanova) [inline Task_3/Tests/test_Min_Max_tupl.py:5]: Could you unite all positive test cases in 1 test by using `@pytest.mark.parametrize`?

File: Task_3/data1.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+1
```
Comment (annkupriyanova) [inline Task_3/data1.txt:1]: This is the right file for testing: 1 number per line

File: Task_4/Check_sum_of_four.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
+def check_sum_of_four(a, b, c, d):
+    quantity = 0
+    dict_sum_a_b = {}
```
Comment (annkupriyanova) [inline Task_4/Check_sum_of_four.py:3]: It's better to use `defaultdict(int)` here to get rid of the check like in line 7

File: Task_4/Check_sum_of_four.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+            else:
+                dict_sum_a_b[i+j] = 1
+    for k in c:
+        for n in d:
+            if k+n in dict_sum_c_d:
+                dict_sum_c_d[k+n] += 1
+            else:
+                dict_sum_c_d[k+n] = 1
+    for key in dict_sum_c_d:
+        if key in dict_sum_a_b:
```
Comment (annkupriyanova) [inline Task_4/Check_sum_of_four.py:18]: Shouldn't it check for `-key`?
I guess the function works incorrectly. Try it with only positive numbers, e.g. [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1], [1, 1, 1, 1, 1]

File: Task_4/Test/test_Sum_of _fourr.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
+from Homework_01.Task_4.Check_sum_of_four import check_sum_of_four
+
+
+def test_sum():
```
Comment (annkupriyanova) [inline Task_4/Test/test_Sum_of _fourr.py:4]: It's recommended to check with different test data to reveal possible errors

File: Task_5/Test/test_maximal_subarray_summ.py [hunk 11]
```diff
@@ -0,0 +1,14 @@
+from Homework_01.Task_5.find_maximal_subarray_sum import find_maximal_subarray_sum
+
+
+def test_max_sum():
+    assert find_maximal_subarray_sum([1, 3, -1, -3, 5, 3, 6, 7], 3) == 16
+
+
+def test_max_sum_for_k0():
+    assert find_maximal_subarray_sum([1, 3, -1, -3, 5, 3, 6, 7], 0) == 0
```
Comment (annkupriyanova) [inline Task_5/Test/test_maximal_subarray_summ.py:9]: It's better to check input parameters for validity. We shouldn't process the input if k=0, better raise exception

File: Task_5/Test/test_maximal_subarray_summ.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def test_max_sum_fot_empty_list():
+    assert find_maximal_subarray_sum([], 3) == 0
```
Comment (annkupriyanova) [inline Task_5/Test/test_maximal_subarray_summ.py:13]: The same concern as above

File: Task_5/find_maximal_subarray_sum.py [hunk 13]
```diff
@@ -0,0 +1,7 @@
+def find_maximal_subarray_sum(nums, k):
```
Comment (annkupriyanova) [inline Task_5/find_maximal_subarray_sum.py:1]: Check the logic. The input [1, 3, -1, -3, 5, -3, 6, 7] and k=3 should give the result = 13.

There is a nuance in a task: "You need to find a sub-array with length **less [or] equal** to "k", with maximal sum."

**Review by annkupriyanova — CHANGES_REQUESTED** [review]
